blog/deployment.yaml

```diff
@@ -0,0 +1,72 @@
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: blog
+  labels:
+    app: blog
+spec:
+  serviceName: blog
+  replicas: 1
+  selector:
+    matchLabels:
+      app: blog
+  template:
+    metadata:
+      labels:
+        app: blog
+    spec:
+{% if deploy %}
+      priorityClassName: production
+{% endif %}
+      containers:
+      - name: blog
+        image: ghost:3.0-alpine
+        env:
+          - name: NODE_ENV
+            value: production
+          - name: url
+{% if deploy %}
+            value: http://blog.hail.is
```

Comment:
I think these should be https

Reply:
Since we're handling the ssl/certificate stuff at the gateway level and using entirely http behind the gateway, I think these have to be http connections to work with our nginx setup. The only annoyance is that links on the website will default to e.g. `http://blog.hail.is/post/foo` which will redirect to https once they hit our gateway, but I don't think we can/want to have the ghost server think it should be using https.

Reply:
(setting the url to be https disables unsecured http connections, which we want to be using within our cluster.)

Reply:
I think that https should work fine. It will be stripped in gateway (~line 42) inside of proxy pass. Gateway also has a server block that listens on port 80 and redirects all requests to port 443/https with a 301 redirect

```
server {
    listen 80 default_server;
    listen [::]:80 default_server;
    server_name _;

    location /.well-known/acme-challenge {
        proxy_pass http://letsencrypt;

        proxy_set_header Host $http_host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host $http_host;
        proxy_set_header X-Forwarded-Proto https;
    }

    location / {
        return 301 https://$host$request_uri;
    }
}
```

Reply:
(Since `https://blog.hail.is` would be the public url / what people click on)

Reply:
The problem with telling ghost that it has an https address isn't that https doesn't work externally (it does) but what ghost does with that information. If I specify the URL as an https URL, the Ghost server won't allow normal http connections at all, which I don't believe is the desired behavior. 

I think I want all incoming external traffic to be redirected to https when it hits our gateway, and I want all of our internal traffic behind the gateway (within our cluster, behind the ssl layer) to be plain http. if I tell Ghost that its address is https, I can (at best) set up a redirect for all incoming http traffic that hits the server to go to https, but I can't tell it to also permit requests via http. I don't think it's necessary to set up the https redirect at the Ghost server level, and it also messes with the liveness/readiness probe setup---if I try to send a http get request to a ghost server with https enabled, I'll get a 301 response instead of a 200, which fails the health check.

Reply:
The unfortunate side effect is that Ghost now thinks it should use http://blog.hail.is/ when creating links, but that's redirected when the traffic from those links hits the gateway, so I think that's probably fine.

Reply:
(it's kind of annoying that specifying the URL informs so much of ghost's default behavior when all I really want is an option that says "here's the prefix you should use when linking to yourself, please trust me to handle the rest", but afaict we don't really get more granularity in configuration)

Reply:
I think Alex is referencing the solution here: https://forum.ghost.org/t/ssl-configured-ghost-generates-a-few-non-https-links-in-the-body/5888/6, if we specify `x-forwarded-proto` then Ghost will allow non-SSL connections.

Reply:
I don't see how this helps me with the healthcheck issue, which afaict doesn't go through the nginx server? Or were you suggesting something else?

Reply:
oh wait hmmm

Reply:
I can set the headers in the probes as well, apparently, which should hopefully fix the issue there.

So if I'm understanding this correctly, I want to have the router's nginx config for the blog.* also set that header, right? So that'll ensure that all traffic that hits a Ghost server that I've deployed in any namespace will have X-Forwarded-Proto=https in the header which will prevent the redirect?

Reply:
@danking

Reply:
ok, cool. I think I set it up correctly and everything *appears* to be working with the https url. I don't think there's any difference in configuration between the internally deployed configuration and the configuration that blog.hail.is would use that I need to account for, right?

Reply:
Yep, I think this is correct. Only relevant differences are that Host will be 1) $service.internal, and that the path will have 2) /namespace/service. You've accounted for both!

Reply:
What does it mean for the host to be $service.internal here? I don't quite understand where that's used.

Reply:
When gateway receives a request to internal.hail.is it matches on a regex pattern (something like /([^\/]+)/([^\/]+) ), that corresponds to /namespace/service.

It thens sets the proxied request's Host field to $service.internal (in your case it would be blog.internal).

Reply:
right---I guess my question is why we're setting the Host field to e.g. `blog.internal` here---what does that do?

Reply:
Creates an internal hostname that the internal service could listen on (blog.internal).

Reply:
Nginx uses the Host field to determine match server_name http://nginx.org/en/docs/http/request_processing.html

Reply:
oooh, that makes sense. Would it be reasonable to just omit setting the `Host` header? I see several other servers blocks don't do it.

Reply:
Yeah I don't think we need it here, since Ghost won't use it when X-Forwarded-Host is set, as long as it trusts the upstream client, which it does by default.